### PR TITLE
upgraded haiku guide to include NNX

### PR DESCRIPTION
**Pull request by chiamp**
added a [Haiku/Linen vs NNX comparison guide](https://flax--3923.org.readthedocs.build/en/3923/nnx/haiku_linen_vs_nnx.html) (based off of our [Haiku to Linen upgrade guide](https://flax.readthedocs.io/en/latest/guides/converting_and_upgrading/haiku_migration_guide.html)).

### CODE REVIEW

File: docs/experimental/nnx/haiku_linen_vs_nnx.rst [hunk 1]
```diff
@@ -0,0 +1,1386 @@
+
+Haiku and Linen vs NNX
```
Comment (cgarciae) [inline docs/experimental/nnx/haiku_linen_vs_nnx.rst:2]: I think we should call the "Migrating from Haiku/Linen to NNX"

**Review by cgarciae — CHANGES_REQUESTED** [review]

File: docs/experimental/nnx/haiku_linen_vs_nnx.rst [hunk 2]
```diff
@@ -0,0 +1,1386 @@
...
+============================
+
+This guide will showcase the differences between Haiku, Flax Linen and Flax NNX.
+Both Haiku and Flax Linen enforce a functional paradigm with stateless modules,
+while NNX is a new, next-generation API that introduces methods that allow the
+user to switch between stateful and stateless modules.
```
[the remark this answers is not in the document]
Comment (chiamp) [reply docs/experimental/nnx/haiku_linen_vs_nnx.rst:8]: I'll also change the tabs to Linen and NNX

File: docs/experimental/nnx/haiku_linen_vs_nnx.rst [hunk 3]
```diff
@@ -0,0 +1,1386 @@
...
+
+      )
+      return optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
+
+    grads = nnx.grad(loss_fn)(model)
+    # we can use Ellipsis to filter out the rest of the variables
+    _, params, _ = nnx.split(model, nnx.Param, ...)
+    params = jax.tree_util.tree_map(lambda p, g: p - 0.1 * g, params, grads)
+
+    nnx.update(model, params)
```
[the remark this answers is not in the document]
Comment (chiamp) [reply docs/experimental/nnx/haiku_linen_vs_nnx.rst:372]: change this for both the dropout and batch norm example

File: docs/experimental/nnx/haiku_linen_vs_nnx.rst [hunk 4]
```diff
@@ -0,0 +1,1386 @@
...
+
+  def train_step(model, inputs, labels):
+    def loss_fn(model):
+      model.train() # set deterministic=False
+      logits = model(
+
+        inputs, # <== inputs
+
+      )
+      return optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
```
[the remark this answers is not in the document]
Comment (chiamp) [reply docs/experimental/nnx/haiku_linen_vs_nnx.rst:365]: do this for all train steps

File: docs/experimental/nnx/haiku_linen_vs_nnx.rst [hunk 5]
```diff
@@ -0,0 +1,1641 @@
...
+      loss = optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
+      return loss, (graphdef, new_state)
+
+    grads, (graphdef, new_state) = jax.grad(loss_fn, has_aux=True)(state.params, state.batch_stats)
+
+    _, _, batch_stats = nnx.split(nnx.merge(graphdef, new_state), nnx.Param, nnx.BatchStat)
+    state = state.apply_gradients(grads=grads)
+    state = state.replace(batch_stats=batch_stats)
+
+    return state
```
Comment (cgarciae) [inline docs/experimental/nnx/haiku_linen_vs_nnx.rst:850]:
```suggestion
  def train_step(state, inputs, labels):
    def loss_fn(params, batch_stats):
      logits, (graphdef, new_state) = state.apply_fn(
        params, batch_stats


      )(inputs) # <== inputs
      _, batch_stats = new_state.split(nnx.Params, nnx.BatchStats)
      loss = optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
      return loss, batch_stats

    grads, batch_stats = jax.grad(loss_fn, has_aux=True)(state.params, state.batch_stats)

    state = state.apply_gradients(grads=grads)
    state = state.replace(batch_stats=batch_stats)

    return state
```

**Review by cgarciae — APPROVED** [review]
